libraries/AP_Scheduler/AP_Scheduler.tpp

```diff
@@ -0,0 +1,213 @@
+template <class Vehicle>
+AP_Scheduler<Vehicle>::~AP_Scheduler() {
+    if(_last_run) {
+        delete [] _last_run;
+    }
+    if(_perf_counters) {
+        delete [] _perf_counters;
+    }
+}
+
+// initialise the scheduler
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::init(const AP_Task<Vehicle> *tasks, uint8_t num_tasks)
+{
+    if(tasks == nullptr || num_tasks == 0) {
+        return;
+    }
+    _tasks = tasks;
+    _num_tasks = num_tasks;
+    _last_run = new uint16_t[_num_tasks];
+    memset(_last_run, 0, sizeof(_last_run[0]) * _num_tasks);
+    _tick_counter = 0;
+}
+
+// one tick has passed
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::tick(void)
+{
+    _tick_counter++;
+}
+
+/*
+  run one tick
+  this will run as many scheduler tasks as we can in the specified time
+ */
+template <class Vehicle>
+void AP_Scheduler<Vehicle>::run(uint32_t time_available)
+{
+    // break condition
+    if(_tasks == nullptr) {
+        return;
+    }
+  
+    if (_debug > 3 && _perf_counters == nullptr) {
+        _perf_counters = new AP_HAL::Util::perf_counter_t[_num_tasks];
+        for (uint8_t i=0; i<_num_tasks && _perf_counters != nullptr; i++) {
+            _perf_counters[i] = hal.util->perf_alloc(AP_HAL::Util::PC_ELAPSED, _tasks[i]._name);
+        }
+    }
+
+    for (uint8_t i=0; i<_num_tasks && _tasks[i]._max_time_micros <= time_available; i++) {
+        const uint16_t dt = _tick_counter - _last_run[i];
+        const uint16_t interval_ticks = (_loop_rate_hz / _tasks[i]._rate_hz) < 1 ? 1 : (_loop_rate_hz / _tasks[i]._rate_hz);
+                
+        if(dt < interval_ticks) {
+            continue;
+        }
+    
+        // run it
+        _task_time_started = AP_HAL::micros();
+        current_task = i;
+        if (_debug > 3 && _perf_counters[i]) {
+            hal.util->perf_begin(_perf_counters[i]);
+        }
+
+        // call that function :)
+        if(_parent) {
+            task_fn_t func = _tasks[i]._function;
+            (_parent->*func)();
+        }
+        
+        if (_debug > 3 && _perf_counters[i]) {
+            hal.util->perf_end(_perf_counters[i]);
+        }
+        current_task = -1;
+        
+        // record the tick counter when we ran. This drives
+        // when we next run the event
+        _last_run[i] = _tick_counter;
+
+        // work out how long the event actually took
+        const uint32_t time_taken = AP_HAL::micros() - _task_time_started;
+
+        // we've slipped a whole run of this task!
+        if (dt >= interval_ticks*2 && _debug > 1) {
+            print_task_info("Scheduler slip task", i, dt, interval_ticks);
+        }
+        // the event overran!
+        if (time_taken > _tasks[i]._max_time_micros && _debug > 4) {
+            print_task_info("Scheduler overrun task", i, dt, time_taken);
+        }
+
+        if (time_taken >= time_available) {
+            update_spare_ticks(0);
+            return;
+        }
+        time_available -= time_taken;
```

Comment:
It's not obvious, but the sum of time_taken is not the time it took. Therefore you can't use it to subtract from time_available.  This is particularly troublesome if the main thread is not the one with the highest priority or you have high irq rate. Both are true on PX4 and Linux (on the latter this is actually something to be fixed).  The main thread could be interrupted (low error) or scheduled out (possibly high error).

This is something we can discuss outside the scope of this PR though as the current code has the same problem.